### PR TITLE
Add support for external stopping in loops

### PR DESCRIPTION
**Pull request by ananthsub**
Summary:
Currently the loops only stop based on the progress information. This can be limiting in case users want to stop at arbitrary points during the step functions. a common use case is early stopping based on a monitored metric.

This adds a method to the State object to indicate the loops should stop. In distributed training, it's contingent on users to synchronize the decision of whether to call `stop()`.

To help with this, under torchtnt/utils we offer this helper function that can meet this need: https://github.com/pytorch/tnt/blob/0feb0c1a50297dab99729410ec332f964143e1e5/torchtnt/utils/distributed.py#L289-L292

Differential Revision: D39529183

### CODE REVIEW

**Review by edward-io — APPROVED** [review]
LGTM
